File: android/src/main/java/app/capgo/capacitor/alarm/CapgoAlarmPlugin.java
package app.capgo.capacitor.alarm;

import android.app.AlarmManager;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.provider.AlarmClock;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapgoAlarm")
public class CapgoAlarmPlugin extends Plugin {

    private final String pluginVersion = "8.0.6";
    private static final String TAG = "CapgoAlarmPlugin";

    // ===== Native OS Alarm helpers (Android) =====
    @PluginMethod
    public void createAlarm(PluginCall call) {
        Integer hour = call.getInt("hour");
        Integer minute = call.getInt("minute");
        if (hour == null || minute == null) {
            call.reject("hour and minute are required");
            return;
        }
        String label = call.getString("label");
        boolean skipUi = call.getBoolean("skipUi", false);
        boolean vibrate = call.getBoolean("vibrate", false);

        Intent intent = new Intent(AlarmClock.ACTION_SET_ALARM)
            .putExtra(AlarmClock.EXTRA_HOUR, hour)
            .putExtra(AlarmClock.EXTRA_MINUTES, minute)
            .putExtra(AlarmClock.EXTRA_VIBRATE, vibrate);
        if (label != null) intent.putExtra(AlarmClock.EXTRA_MESSAGE, label);
        if (skipUi) intent.putExtra(AlarmClock.EXTRA_SKIP_UI, true);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        try {
            getContext().startActivity(intent);
            JSObject ret = new JSObject();
            ret.put("success", true);
            call.resolve(ret);
        } catch (Exception e) {
            JSObject ret = new JSObject();
            ret.put("success", false);
            ret.put("message", e.getMessage());
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void openAlarms(PluginCall call) {
        Intent intent = new Intent(AlarmClock.ACTION_SHOW_ALARMS);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        try {
            getContext().startActivity(intent);
            JSObject ret = new JSObject();
            ret.put("success", true);
            call.resolve(ret);
        } catch (Exception e) {
            JSObject ret = new JSObject();
            ret.put("success", false);
            ret.put("message", e.getMessage());
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void getOSInfo(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("platform", "android");
        ret.put("version", Build.VERSION.RELEASE);
        ret.put("supportsNativeAlarms", true);
        ret.put("supportsScheduledNotifications", true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            AlarmManager am = (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);
            boolean can = am != null && am.canScheduleExactAlarms();
            ret.put("canScheduleExactAlarms", can);
        }
        call.resolve(ret);
    }

    @Override
    public void requestPermissions(PluginCall call) {
        boolean requestExact = call.getBoolean("exactAlarm", false);
        JSObject ret = new JSObject();
        if (requestExact && Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            AlarmManager am = (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);
            boolean can = am != null && am.canScheduleExactAlarms();
            if (!can) {
                Intent i = new Intent(android.provider.Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM);
                i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                try {
                    getContext().startActivity(i);
                } catch (Exception ignored) {}
                ret.put("granted", false);
                JSObject details = new JSObject();
                details.put("exactAlarm", false);
                ret.put("details", details);
                call.resolve(ret);
                return;
            }
        }
        ret.put("granted", true);
        call.resolve(ret);
    }

    @Override
    public void checkPermissions(PluginCall call) {
        JSObject ret = new JSObject();
        JSObject details = new JSObject();
        boolean granted = Build.VERSION.SDK_INT < Build.VERSION_CODES.S;
        boolean hasDetails = false;
        String message = null;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            try {
                Context context = getContext();
                if (context == null) {
                    throw new IllegalStateException("Plugin context unavailable");
                }
                AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
                boolean canExact = am != null && am.canScheduleExactAlarms();
                details.put("exactAlarm", canExact);
                hasDetails = true;
                granted = canExact;
                if (!canExact) {
                    message = "Exact alarm permission not granted";
                }
            } catch (Exception e) {
                details.put("exactAlarm", false);
                hasDetails = true;
                message = "Failed to query exact alarm capability";
                Log.w(TAG, "Unable to determine exact alarm capability status", e);
                granted = false;
            }
        } else {
            message = "Exact alarm capability check requires Android S+";
        }

        ret.put("granted", granted);
        if (hasDetails) {
            ret.put("details", details);
        }
        if (message != null) {
            ret.put("message", message);
        }
        call.resolve(ret);
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }

    @PluginMethod
    public void getAlarms(PluginCall call) {
        // Android does not provide a public API to query system alarms
        // Each app must track its own alarms
        JSObject ret = new JSObject();
        ret.put("alarms", new org.json.JSONArray());
        call.resolve(ret);
    }

    @PluginMethod
    public void cancelAlarm(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("success", false);
        ret.put("message", "Cancel alarm is not supported on Android");
        call.resolve(ret);
    }
}

File: ios/Sources/CapgoAlarmPlugin/CapgoAlarmPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapgoAlarmPlugin)
public class CapgoAlarmPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.0.6"
    public let identifier = "CapgoAlarmPlugin"
    public let jsName = "CapgoAlarm"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "createAlarm", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAlarms", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getOSInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAlarms", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelAlarm", returnType: CAPPluginReturnPromise)
    ]

    @objc func createAlarm(_ call: CAPPluginCall) {
        let hour = call.getInt("hour") ?? -1
        let minute = call.getInt("minute") ?? -1
        if hour < 0 || minute < 0 { call.reject("hour and minute are required"); return }
        let label = call.getString("label")

        AlarmKitBridge.createAlarm(hour: hour, minute: minute, label: label) { success, message, alarmId in
            var result: [String: Any] = ["success": success, "message": message ?? NSNull()]
            if let alarmId = alarmId {
                result["id"] = alarmId
            }
            call.resolve(result)
        }
    }

    @objc func openAlarms(_ call: CAPPluginCall) {
        AlarmKitBridge.openAlarms { success, message in
            call.resolve(["success": success, "message": message ?? NSNull()])
        }
    }

    @objc func getOSInfo(_ call: CAPPluginCall) {
        let version = UIDevice.current.systemVersion
        let supportsNative = AlarmKitBridge.isAvailable()
        call.resolve([
            "platform": "ios",
            "version": version,
            "supportsNativeAlarms": supportsNative,
            "supportsScheduledNotifications": true
        ])
    }

    // Capacitor permission lifecycle
    override public func checkPermissions(_ call: CAPPluginCall) {
        guard AlarmKitBridge.isAvailable() else {
            call.resolve([
                "granted": false,
                "details": ["alarmKit": false],
                "message": "AlarmKit not available on this device/SDK",
            ])
            return
        }

        AlarmKitBridge.currentAuthorizationStatus { granted, statusDescription in
            var result: [String: Any] = [
                "granted": granted,
                "details": ["alarmKit": granted],
            ]
            if !granted, let statusDescription = statusDescription {
                result["message"] = "AlarmKit authorization status: \(statusDescription)"
            }
            call.resolve(result)
        }
    }

    override public func requestPermissions(_ call: CAPPluginCall) {
        // Request AlarmKit authorization when available
        if !AlarmKitBridge.isAvailable() {
            call.resolve(["granted": false])
            return
        }
        AlarmKitBridge.requestAuthorization { granted, message in
            var result: [String: Any] = ["granted": granted]
            if let message = message { result["message"] = message }
            call.resolve(result)
        }
    }

    // No data conversion helpers needed for native-only operations

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

    @objc func getAlarms(_ call: CAPPluginCall) {
        AlarmKitBridge.getAlarms { alarms, error in
            if let error = error {
                // Return empty array with error message for consistency
                call.resolve(["alarms": [], "message": error])
            } else {
                call.resolve(["alarms": alarms])
            }
        }
    }

    @objc func cancelAlarm(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("Missing alarm id")
            return
        }
        AlarmKitBridge.cancelAlarm(id: id) { success, message in
            call.resolve(["success": success, "message": message ?? NSNull()])
        }
    }

}
